src/Atmos/Model/prog_prim_conversion.jl

```diff
@@ -74,11 +76,12 @@ function prognostic_to_primitive!(
     prim::Vars,
     prog::Vars,
     e_int::AbstractFloat,
+    aux::Vars,
 )
-    ts = PhaseDry(parameter_set(atmos), e_int, prog.ρ)
-    prim.ρ = prog.ρ
-    prim.u = prog.ρu ./ prog.ρ
-    prim.p = air_pressure(ts)
+    ts = recover_thermo_state(atmos, prog, aux)
+    prim.ρ = air_density(ts) # Needed for recovery of energy
```

Comment:
If this is called on the boundary (@Zhengyu-Huang can you confirm?), I would recommend calling `new_thermo_state` instead (it's equivalent same at the moment).